File: tools/slice_ram_buttons.py
#!/usr/bin/env python3
"""Slice Ram Buttons sprite sheet into individual transparent PNGs and manifest.

Requirements:
  pip install pillow

Run:
  python tools/slice_ram_buttons.py

By default it expects the source sheet at assets/ui/Ram Buttons.png
If your file currently lives at assets/botones/Ram Buttons.png adjust INPUT below
or pass an env var: RAM_SHEET=assets/botones/"Ram Buttons.png" python tools/slice_ram_buttons.py
"""
from __future__ import annotations
from PIL import Image
import json, os, sys

INPUT = os.environ.get("RAM_SHEET", "assets/ui/Ram Buttons.png")
OUTDIR = "assets/ui/ram"
BTN_DIR = os.path.join(OUTDIR, "buttons")
ICO_DIR = os.path.join(OUTDIR, "icons")
MANIFEST = os.path.join(OUTDIR, "manifest.json")

MIN_PIXELS = 12        # component area threshold to skip noise
BUTTON_SIZE_THRESHOLD = 40  # classify as button if max(width,height) >= this
ROW_BUCKET_H = 40      # height bucket for ordering (group components into rows)

def main():
    if not os.path.isfile(INPUT):
        alt = "assets/botones/Ram Buttons.png"
        if os.path.isfile(alt):
            print(f"[info] INPUT not found, using fallback {alt}")
            global INPUT
            INPUT = alt
        else:
            print(f"Error: sheet not found: {INPUT}")
            return 2

    os.makedirs(BTN_DIR, exist_ok=True)
    os.makedirs(ICO_DIR, exist_ok=True)

    im = Image.open(INPUT).convert("RGBA")
    W, H = im.size
    px = im.load()

    # Binary mask: True if NOT pure black (#000000)
    mask = [[(px[x, y][0] | px[x, y][1] | px[x, y][2]) != 0 for x in range(W)] for y in range(H)]
    visited = [[False]*W for _ in range(H)]

    sys.setrecursionlimit(1000000)

    def flood(y0:int, x0:int):
        stack=[(y0,x0)]
        visited[y0][x0]=True
        minx=maxx=x0
        miny=maxy=y0
        count=0
        while stack:
            y,x = stack.pop()
            count += 1
            if x<minx: minx=x
            if x>maxx: maxx=x
            if y<miny: miny=y
            if y>maxy: maxy=y
            for dy,dx in ((1,0),(-1,0),(0,1),(0,-1)):
                ny, nx = y+dy, x+dx
                if 0<=ny<H and 0<=nx<W and not visited[ny][nx] and mask[ny][nx]:
                    visited[ny][nx]=True
                    stack.append((ny,nx))
        # expand 1px margin
        minx = max(0, minx-1); miny = max(0, miny-1)
        maxx = min(W-1, maxx+1); maxy = min(H-1, maxy+1)
        return (minx, miny, maxx+1, maxy+1, count)

    boxes=[]
    for y in range(H):
        for x in range(W):
            if mask[y][x] and not visited[y][x]:
                minx,miny,maxx,maxy,count = flood(y,x)
                if count >= MIN_PIXELS:
                    boxes.append((minx,miny,maxx,maxy,count))

    if not boxes:
        print("No components detected; aborting.")
        return 1

    # sort into row buckets (top->down) and left->right
    boxes.sort(key=lambda b:(b[1]//ROW_BUCKET_H, b[0]))
    rows={}
    for b in boxes:
        r = b[1]//ROW_BUCKET_H
        rows.setdefault(r,[]).append(b)

    manifest={"buttons":[], "icons":[]}
    row_idx=0
    for r in sorted(rows.keys()):
        col_idx=0
        for (x0,y0,x1,y1,_cnt) in rows[r]:
            w = x1-x0; h = y1-y0
            crop = im.crop((x0,y0,x1,y1)).copy()
            # convert black to transparent
            data = crop.getdata()
            new = []
            for (R,G,B,A) in data:
                if R==0 and G==0 and B==0:
                    new.append((0,0,0,0))
                else:
                    new.append((R,G,B,A))
            crop.putdata(new)
            is_button = max(w,h) >= BUTTON_SIZE_THRESHOLD
            name = f"{'btn' if is_button else 'ico'}_r{row_idx:02d}_c{col_idx:02d}.png"
            outdir = BTN_DIR if is_button else ICO_DIR
            outpath = os.path.join(outdir, name)
            crop.save(outpath, "PNG", optimize=True)
            rec = {"name":name, "w":w, "h":h, "path":f"assets/ui/ram/{'buttons' if is_button else 'icons'}/{name}"}
            (manifest["buttons"] if is_button else manifest["icons"]).append(rec)
            col_idx += 1
        row_idx += 1

    with open(MANIFEST, "w", encoding="utf-8") as f:
        json.dump(manifest, f, indent=2, ensure_ascii=False)

    print(f"OK ✓  Exportados {len(manifest['buttons'])} botones y {len(manifest['icons'])} iconos.")
    print(f"Manifest: {MANIFEST}")

if __name__ == "__main__":
    raise SystemExit(main())

File: tools/slice_ram_buttons.mjs
#!/usr/bin/env node
/**
 * Node/Sharp variant of Ram Buttons slicer.
 * Requires: npm i sharp
 * Run: node tools/slice_ram_buttons.mjs
 */
import fs from 'node:fs';
import sharp from 'sharp';

let INPUT = process.env.RAM_SHEET || 'assets/ui/Ram Buttons.png';
if (!fs.existsSync(INPUT)) {
  const alt = 'assets/botones/Ram Buttons.png';
  if (fs.existsSync(alt)) { console.log('[info] Using fallback', alt); INPUT = alt; }
}
const OUTDIR = 'assets/ui/ram';
const BTN = `${OUTDIR}/buttons`; const ICO = `${OUTDIR}/icons`;
fs.mkdirSync(BTN, {recursive:true}); fs.mkdirSync(ICO, {recursive:true});

const MIN_PIXELS = 12;
const BUTTON_SIZE_THRESHOLD = 40;
const ROW_BUCKET_H = 40;

const img = sharp(INPUT).ensureAlpha();
const meta = await img.metadata();
const W = meta.width, H = meta.height;
const raw = await img.raw().toBuffer(); // RGBA

const idx=(x,y)=> (y*W + x)*4;
const isNonBlack=(x,y)=> { const i=idx(x,y); return raw[i]||raw[i+1]||raw[i+2]; };
const visited=new Uint8Array(W*H);

function flood(sx,sy){
  const stack=[sx,sy]; visited[sy*W+sx]=1;
  let minx=sx,maxx=sx,miny=sy,maxy=sy,count=0;
  while(stack.length){
    const y=stack.pop(), x=stack.pop(); count++;
    if(x<minx)minx=x; if(x>maxx)maxx=x; if(y<miny)miny=y; if(y>maxy)maxy=y;
    const nb=[x+1,y, x-1,y, x,y+1, x,y-1];
    for(let i=0;i<nb.length;i+=2){
      const nx=nb[i], ny=nb[i+1];
      if(nx>=0&&ny>=0&&nx<W&&ny<H && !visited[ny*W+nx] && isNonBlack(nx,ny)){
        visited[ny*W+nx]=1; stack.push(ny,nx);
      }
    }
  }
  // expand 1px
  minx=Math.max(0,minx-1); miny=Math.max(0,miny-1);
  maxx=Math.min(W-1,maxx+1); maxy=Math.min(H-1,maxy+1);
  return {minx,miny,maxx,maxy,count};
}

const boxes=[];
for(let y=0;y<H;y++){
  for(let x=0;x<W;x++){
    if(!visited[y*W+x] && isNonBlack(x,y)){
      const b=flood(x,y); if(b.count>=MIN_PIXELS) boxes.push(b);
    }
  }
}
if(!boxes.length){ console.error('No components detected'); process.exit(1); }

boxes.sort((a,b)=> (Math.floor(a.miny/ROW_BUCKET_H)-Math.floor(b.miny/ROW_BUCKET_H)) || (a.minx-b.minx));
const rowMap={};
for(const b of boxes){
  const r=Math.floor(b.miny/ROW_BUCKET_H); (rowMap[r] ||= []).push(b);
}

const manifest={ buttons:[], icons:[] };
let rowIdx=0;
for(const r of Object.keys(rowMap).map(Number).sort((a,b)=>a-b)){
  let colIdx=0;
  for(const b of rowMap[r]){
    const w=b.maxx-b.minx+1, h=b.maxy-b.miny+1;
    const extract = await sharp(INPUT).extract({ left:b.minx, top:b.miny, width:w, height:h }).ensureAlpha();
    const buf = await extract.raw().toBuffer();
    // black -> transparent
    for (let i=0;i<buf.length;i+=4){ if(!buf[i] && !buf[i+1] && !buf[i+2]) buf[i+3]=0; }
    const isBtn = Math.max(w,h) >= BUTTON_SIZE_THRESHOLD;
    const name = `${isBtn?'btn':'ico'}_r${String(rowIdx).padStart(2,'0')}_c${String(colIdx).padStart(2,'0')}.png`;
    const outPath = `${isBtn?BTN:ICO}/${name}`;
    await sharp(buf,{raw:{width:w,height:h,channels:4}}).png({compressionLevel:9}).toFile(outPath);
    const rec={ name, w, h, path:`assets/ui/ram/${isBtn?'buttons':'icons'}/${name}` };
    (isBtn?manifest.buttons:manifest.icons).push(rec);
    colIdx++;
  }
  rowIdx++;
}

fs.writeFileSync(`${OUTDIR}/manifest.json`, JSON.stringify(manifest,null,2), 'utf8');
console.log(`OK ✓ Exportados ${manifest.buttons.length} botones y ${manifest.icons.length} iconos.`);
console.log('Manifest: assets/ui/ram/manifest.json');
